_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+There were some really nasty corner cases, such as hash tables that 'worked', but ended up putting every object in the first hash bucket, giving truly awful performance.
+
+Alex and Jess fixed that in the compiler, but the core hardware abstractions remained largely unchanged.
+In CHERIoT, we removed the last vestiges of offsets by removing the set-offset instruction, because the extra addition hurt critical path lengths.
+I still believe that this was the right thing to try for research, but it's not research if you know it will work in advance, and this one didn't work at all.
+
+The other big change that was in before the paper was to generalise the sealing mechanism.
+The [first public version of the CHERI ISA](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-850.pdf) contained two instructions for sealing, `CSealCode` and `CSealData` (pages 97 and 99).
+These were intended to create cross-domain entry points.
+A cross-compartment call would be represented by two capabilities sealed with the same type.
+These would be installed in the PCC and an invoked-data capability register during a compartment call.
+
+By the 2015 paper, these were replaced by a single `CSeal` instruction that *could* seal pairs of capabilities in this way but could also seal capabilities independently.
+This let us implement opaque data types using sealed capabilities, something that we do extensively in CHERIoT (see [this section of the CHERIoT book for a longer explanation of CHERI object types and sealing](https://cheriot.org/book/concepts.html#sealing_intro)).
+I think our first public use of this (it was used a lot as an aside elsewhere) was our [2017 paper that extended type safety from Java code across the JNI boundary](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf).
+
+The `CCall` instruction at the time was actually a trap (with the expectation that a future version would move some or all of the trap handler into hardware).
+It branched to a handler, as did `CReturn`.
+The software stack in the paper used these to implement a trusted stack for cross-compartment calls.
+In CHERIoT, we use a very similar model but don't need a special instruction.
+We use the sealed entry (sentry) model, where one object type is reserved for capabilities that you can branch to but can't do anything else with.
+This was added a few years after the 2015 paper and meant that CHERIoT started from a model where the trap was not necessary.
+Using sentries means that our cross-compartment call can just use a jump to a sentry to invoke the switcher, rather than a trap.
+This is more scalable: a single system can provide multiple switchers like ours, but the 2015 paper required one per address space.
+In CHERIoT we extend sentries to control interrupt state and to differentiate between forward and backwards control-flow edges.
```

Comment:
I’m not sure whether it’s worth commenting on or not, but the same sealed entry capability mechanism is used for all control-flow pointers in CheriBSD/CheriABI following another Microsoft interaction with the University project: Saar Amar’s observation that, when used, this disrupted exploitation of intentionally injected vulnerabilities around uninitialised memory in WebKit JSC. I’m having trouble tracking down the commit date that enabled sentries by default, though.